pallets/proxy-teleport/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+			beneficiary: Box<VersionedMultiLocation>,
+			native_asset_amount: u128,
+			proxy_asset: Box<VersionedMultiAssets>,
+			fee_asset_item: u32,
+		) -> DispatchResult {
+			Self::do_proxy_teleport_assets(
+				origin,
+				dest,
+				beneficiary,
+				native_asset_amount,
+				proxy_asset,
+				fee_asset_item,
+			)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_proxy_teleport_assets(
+		origin: OriginFor<T>,
+		dest: Box<VersionedMultiLocation>,
+		beneficiary: Box<VersionedMultiLocation>,
+		native_asset_amount: u128,
+		proxy_asset: Box<VersionedMultiAssets>,
+		fee_asset_item: u32,
+	) -> DispatchResult {
+		//Unbox origin, destination and beneficiary.
+		let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin)?;
+		let dest: MultiLocation =
+			(*dest).try_into().map_err(|()| pallet_xcm::Error::<T>::BadVersion)?;
+		let beneficiary: MultiLocation =
+			(*beneficiary).try_into().map_err(|()| pallet_xcm::Error::<T>::BadVersion)?;
+
+		//Create assets
+
+		// Native from local perspective
+		let native_asset = MultiAsset {
+			id: AssetId::Concrete(MultiLocation::here()),
+			fun: Fungibility::Fungible(native_asset_amount),
+		};
+		let assets = MultiAssets::from(vec![native_asset.clone()]);
+
+		// Native from foreign perspective
+		let context = T::UniversalLocation::get();
+		let native_as_foreign = native_asset
+			.reanchored(&dest, context)
+			.map_err(|_| pallet_xcm::Error::<T>::CannotReanchor)?;
+		let foreing_assets = MultiAssets::from(vec![native_as_foreign]);
+
+		//Unbox proxy asset
+		let proxy_asset: MultiAssets =
+			(*proxy_asset).try_into().map_err(|()| pallet_xcm::Error::<T>::BadVersion)?;
+
+		//TeleportFilter check
+		let value = (origin_location, assets.into_inner());
+		ensure!(T::XcmTeleportFilter::contains(&value), pallet_xcm::Error::<T>::Filtered);
+		let (origin_location, assets) = value;
+
+		// Reanchor the proxy asset to the destination chain.
+
+		let fees = proxy_asset
+			.get(fee_asset_item as usize)
+			.ok_or(pallet_xcm::Error::<T>::Empty)?
+			.clone()
+			.reanchored(&dest, context)
+			.map_err(|_| pallet_xcm::Error::<T>::CannotReanchor)?;
+
+		// TODO: Define if Withdrawn proxy assets are deposited or trapped.
+		// Check if there is no vulnerability through RefundSurplus
+		//let max_assets = (assets.len() as u32).checked_add(1).ok_or(Error::<T>::TooManyAssets)?;
+
+		//Build the message to execute on origin.
+		let assets: MultiAssets = assets.into();
+		let message: Xcm<<T as frame_system::Config>::RuntimeCall> = Xcm(vec![
+			// Withdraw drops asset so is used as burn mechanism
+			WithdrawAsset(assets.clone()),
+			BurnAsset(assets),
+		]);
+
+		// Build the message to send.
+		// Set WeightLimit
+		// TODO: Implement weight_limit calculation with final instructions.
+		let weight_limit: WeightLimit = Unlimited;
+		let xcm_message: Xcm<()> = Xcm(vec![
+			WithdrawAsset(proxy_asset),
+			BuyExecution { fees, weight_limit },
+			ReceiveTeleportedAsset(foreing_assets.clone()),
```

Comment:
I don't get the need for the `ReceiveTeleportedAsset` instruction .. can you explain ?

Reply:
Hey! Sure. 

We must take in consideration that:
-  `InitiateTeleport` instruction imprints a `ReceiveTeleportedAsset` as first instruction of [the xcm message ](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/xcm-executor/src/lib.rs#L662)that it sends. 
- The `Barrier` that we are targeting for in Asset Hub is `AllowTopLevelPaidExecutionFrom`, this Barrier to pass imposes a certain sequence of Instructions to be received in order as detailed [here](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/xcm-builder/src/barriers.rs#L86). 
- That is why `do_teleport_assets` fn from `pallet-xcm` sets `BuyExecution` as the first instruction of [the xcm message that is sent as parameter](https://github.com/paritytech/polkadot/blob/962bc21352f5f80a580db5a28d05154ede4a9f86/xcm/pallet-xcm/src/lib.rs#L1201) to `InitiateTeleport`, therefore the Instruction order to be receive would be: 
-- `Receive` -> `ClearOrigin` (from `InitiateTeleport`) 
-- `BuyExecution` (from the xcm sent as parameter of `InitiateTeleport`)
Leading to the Barrier compliant sequence of `ReceiveTele...` -> `ClearOrigin` -> `BuyExecution`  

However, on Asset Hub we cannot `BuyExecution` with HOP or teleport ROC into it which would trigger the `ReceiveTeleportedAssets` progression with ROC.

There is no Instruction that allows me to prepend  a `Withdraw (ROC)` + `BuyExecution (ROC)` together with the `InitiateTeleport` logic. 

Therefore I am forced to split the logic of the call in two parts: What should be executed on Trappist and what should be sent to be executed on Asset Hub.  I am trying to handcraft the logic as close as possible to the one from `InitiateTeleport` but adding the extra instructions that are needed for the use case. This leads to:

- A `execute_xcm` for taking the assets to be teleported on origin via `Withdraw` + `Burn` 
- A `send_xcm` for paying execution with ROC and apply the "teleport" logic on Asset Hub. This is done by the `Withdraw` + `BuyExecution` + `ReceiveTele...` + `DepositAsset` that is sent to be executed on Asset Hub.

With these Instructions on this combination I try to approach to the `do_teleport_assets` logic but sneaking an execution buy with ROC on AH. 

I hope this explains both this approach and the rest of the comments, please do let me know if have any further question or if something doesn't make sense since this is just how I figured out how to work around this issue